scripts: west_commands: zspdx: SBOM generation meta-data update

Rico-van-Dongen opened the pull request:
this pull request is a follow-up of discussion:
- https://github.com/zephyrproject-rtos/zephyr/discussions/90604#discussioncomment-13280647

this pull request will update the `PackageName`, `PackageVersionand` and `PackageSupplier` in the generated SBOM based on the information passed in the `external-references` in the module.yml. This way packages are better recognized by vulnerability scanning tools like `cve-bin-tool`.
In addition an extra option `external-references-comment` can now be passed into the module.yml that will add the SPDX `ExternalRefComment` field to the SBOM. These comments can be single line or multi line.

pdgendt commented:
You will need to update your existing commits and force push the changes, use `scripts: west_commands: zspdx:` as the commit subject prefix.

pdgendt commented:
> thanks for your help, much appreciated. I think I fixed it now. Initially I forgot to take the max line length into account but also that should be fixed. However, i did not see the re-run of the compliance check starting yet.

As you are a new contributor, a manual start of the actions has to be done, started them for you

> I tried the local check_compliance.py but that gives me some python errors I need to dig into (probably just missing packages)

Let's see what CI has to say :)

Rico-van-Dongen commented:
ok, also fixed the commit author now but I'm still a bit puzzled by this fail:
Signed-off-by line (Signed-off-by: Rico van Dongen <rcdongen@gmail.com>) does not follow the syntax: First Last <email>.
should I just omit the middle name??

Also, I've seen automated tests running for the west commands. should one of these be added/extended to test my new feature?

Rico-van-Dongen commented:
> There's a simple test in https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/application_development/software_bill_of_materials you might extend that one.

Hi Pieter, I changed the PR to draft as I will need a bit more time to figure out how to run the tests locally first. However, do i understand the current test correctly that it only checks if the spdx command passes and does not check any content of the output files. I guess in order to get full test coverage the dummy project would also need to include some dummy packages with different content in the module.yml. do you have some pointers to similar tests so i can check how to achieve this?

pdgendt commented:
The current test is indeed simply checking that the `west` command succeeds.
I think the simplest approach for adding tests would be to add a python script that parses and verifies the output. You can call that python script from the `ctest` harness.

Rico-van-Dongen commented:
even without additional tests there is already one failing but i doesn't appear to be related to the changes. 
https://github.com/zephyrproject-rtos/zephyr/actions/runs/15436293737/job/43445908162?pr=91013
could this be because my branch is already miles behind main?

pdgendt commented:
> even without additional tests there is already one failing but it doesn't appear to be related to the changes. [zephyrproject-rtos/zephyr/actions/runs/15436293737/job/43445908162?pr=91013](https://github.com/zephyrproject-rtos/zephyr/actions/runs/15436293737/job/43445908162?pr=91013) could this be because my branch is already miles behind main?

That issue is unrelated to this PR, should get fixed soon.

Rico-van-Dongen commented:
I currently have trouble building our project with the latest main so that means I cannot test the merged changes right now. it looks like commit d77b58aacb913d25c30b63620d413eaa94622750 broke something in our set-up as now the linker is unable to locate the toolchain. 
`arm-none-eabi/bin/ld.bfd.exe: cannot find gnu: No such file or directory`

Rico-van-Dongen committed:
scripts: west_commands: zspdx: updating sbom data based on cpe reference

The PackageName, PackageVersionand and PackageSupplier in the generated
SBOM will be updated based on the information passed in the
external-references in the module.yml.
This way packages are better recognized by vulnerability scanning tools
like cve-bin-tool.

Signed-off-by: Rico van Dongen <rdongen@ziggo.nl>

kartben commented on scripts/west_commands/zspdx/datatypes.py:
```diff
@@ -81,9 +81,15 @@ class PackageConfig:
+    # package supplier or vendor
+    supplier: str = ""
+
     # package external references
     externalReferences: list = field(default_factory=list)
 
+    # package external reference Comments
+    externalReferencesComment: list = field(default_factory=list)
+
```
see comment re: schema

kartben commented on scripts/west_commands/zspdx/writer.py:
```diff
@@ -102,14 +112,28 @@ def writePackageSPDX(f, pkg, spdx_version=SPDX_VERSION_2_3):
             f.write(f"ExternalRef: SECURITY cpe23Type {ref}\n")
         elif re.fullmatch(PURL_REGEX, ref):
-            f.write(f"ExternalRef: PACKAGE_MANAGER purl {ref}\n")
+            f.write(f"ExternalRef: PACKAGE-MANAGER purl {ref}\n")
```
this fix should be a different commit please

kartben commented on scripts/west_commands/zspdx/writer.py:
```diff
@@ -102,14 +112,28 @@ def writePackageSPDX(f, pkg, spdx_version=SPDX_VERSION_2_3):
         else:
             log.wrn(f"Unknown external reference ({ref})")
 
+    # external ref comment:
+    if len(pkg.cfg.externalReferencesComment) > 0:
+        for ln, comment in enumerate(pkg.cfg.externalReferencesComment):
+            if ln == 0: # first line
+                f.write("ExternalRefComment: <text>")
+            f.write(f"{comment}")
+            if ln == len(pkg.cfg.externalReferencesComment)-1: # last line
+                f.write("</text>\n")
+            else:
+                f.write("\n")
```
see other comments - this is incorrect, ExternalRefComment are to be serialized next to the ExternalRef they correspond to

kartben reviewed (changes requested):
I sincerely apologize for the time it took me to get to this. Please see some of my comments. Also, in addition to splitting the PACKAGE_MANAGER fix in its own commits, it would be nice to split the externalrefcomment stuff from the rest of the changes. 
Let me know if you have questions. Thanks!

Rico-van-Dongen commented:
@kartben, thanks for your review. As the ExternalRefComment addition is a bit beyond the original purpose of this merge request I'll take that out. It is not being used by any package so it will have limited value anyway. 
About putting the  PACKAGE-MANAGER change into a separate commit, can this still be in this merge request? I thought the idea was to amend and force-push things

Rico-van-Dongen committed:
scripts: west_commands: zspdx: re-naming PACKAGE-MANAGER

According to the SPDX specification
PACKAGE-MANAGER should be spelled with a dash not an underscore

Signed-off-by: Rico van Dongen <rdongen@ziggo.nl>
